src/pages/challenges/get-healthy.js

```diff
@@ -1,34 +1,57 @@
-import Image from 'next/image'
+import styles from '../../styles/Challenges.module.css'
 import HeaderCard from 'components/cards/HeaderCard'
 import ChallengeCategoryList from 'components/challenges/ChallengeCategoryList'
-import styles from '../../styles/Challenges.module.css'
 import DatePrinter from '../../components/layout/DatePrinter'
+import {
+	getTodayTasksByCategory,
+	getTodayScoreByCategory
+} from 'model/helpers/today-tasks'
+import { useState } from 'react'
+import { useUserContext } from 'context/UserContext.js'
+
+export default function GetHealthy({ tasks, progress }) {
+	const [data, setData] = useState({ tasks: tasks, progress: progress })
+	const userId = 1
+	const ctx = useUserContext()
 
-export default function GetHealthy({ tasks }) {
 	const completedHandler = async (taskId) => {
-		const response = await fetch('http//localhost:3000/api/update-status', {
+		const taskPoints = data.tasks.find(
+			(task) => task.id === Number(taskId)
+		).points
+
+		const response = await fetch('http://localhost:3000/api/update-status', {
 			method: 'PUT',
 			headers: {
 				'Content-Type': 'application/json'
 			},
-			body: JSON.stringify({ taskId: 1, date: '2023-03-29' })
+			body: JSON.stringify({ userId: userId, taskId: taskId })
 		})
 
 		if (response.ok) {
-			console.log(`task ${taskId} has been ticked`)
+			setData((prevState) => {
+				const updatedTasks = prevState.tasks.map((task) =>
+					task.id == taskId ? { ...task, status: 1 } : task
+				)
```

Comment:
Not really suggesting you change anything here but think it's a good example to think about data structure choice in a practical (rather than a pointlessly theoretical) way.

If there are loads of `tasks` and the one you want is at the end of the list, you're going to spend a long time mapping through them to get to it and update it. That's fine for small numbers (which is why I don't think you need to change this), but you could do it in a way that doesn't suffer from this problem, and I think is maybe a little easier to understand. It would also work to replace your `find` on line 18! That would be to use an object (rather than an array) of tasks, where the keys are the task IDs:

```typescript
const [data, setData] = useState({
  tasks: tasks.reduce((tasksObject, task) => ({ ...tasksObject, [task.id]: task }), {}),
  progress: progress
});

// ...

setData((prevState) => {
  const tasksCopy = { ...prevState.tasks };
  if (taskId in tasksCopy) {
    tasksCopy[taskId].status = 1;
  }
  // ...
});
```

Reply:
Thanks for this, Jamie! This has been a good learning point for us this morning. We've been exploring its use and will optimise from this suggestion in the coming phase. :)

Reply:
Just coming back to this I would add that some particularly purist pedants might object to the fact that in my example I _mutated_ `tasksCopy`, and they might prefer something _immutable_ like this:
```typescript
const nextTasks = {
  ...prevState.tasks,
  [taskId]: {
    ...prevState.tasks[taskId],
    status: 1
  }
}
```
I have at times been that kind of pedant, but at the moment am not really feeling it - just thought I'd mention the different approaches.